src/BSON/Value.c

```diff
@@ -0,0 +1,631 @@
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_bson_encode.h"
+#include "BSON/Value_arginfo.h"
+#include "BSON/Binary.h"
+#include "BSON/DBPointer.h"
+#include "BSON/Decimal128.h"
+#include "BSON/Document.h"
+#include "BSON/Int64.h"
+#include "BSON/Javascript.h"
+#include "BSON/ObjectId.h"
+#include "BSON/PackedArray.h"
+#include "BSON/Regex.h"
+#include "BSON/Symbol.h"
+#include "BSON/Timestamp.h"
+#include "BSON/UTCDateTime.h"
+
+#define PHONGO_VALUE_TYPE_ISSER(name, type)                              \
+	static PHP_METHOD(MongoDB_BSON_Value, name)                          \
+	{                                                                    \
+		PHONGO_PARSE_PARAMETERS_NONE();                                  \
+		RETURN_BOOL(Z_VALUE_OBJ_P(getThis())->value.value_type == type); \
+	}
+
+#define PHONGO_VALUE_CHECK_TYPE(bson_value, expected_type)                                                       \
+	if (bson_value.value_type != expected_type) {                                                                \
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Unexpected BSON type %d", bson_value.value_type); \
+		return;                                                                                                  \
+	}
+
+zend_class_entry* php_phongo_value_ce;
+
+static HashTable* php_phongo_value_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_value_t* intern;
+	HashTable*          props;
+
+	intern = Z_OBJ_VALUE(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 1);
+
+	// TODO: Print value (somehow)
+	//{
+	//	zval data;
+	//
+	//	ZVAL_STR(&data, php_base64_encode((const unsigned char*) bson_get_data(intern->bson), intern->bson->len));
+	//	zend_hash_str_update(props, "data", sizeof("data") - 1, &data);
+	//}
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_Value)
+
+static PHP_METHOD(MongoDB_BSON_Value, fromPHP)
+{
+	php_phongo_value_t* intern;
+	zval                zv;
+	zval*               data;
+	zend_long           lvalue;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_ZVAL(data);
+	PHONGO_PARSE_PARAMETERS_END();
+
+	object_init_ex(&zv, php_phongo_value_ce);
+	intern = Z_VALUE_OBJ_P(&zv);
+
+	switch (Z_TYPE_P(data)) {
+		case IS_UNDEF:
+		case IS_NULL:
+			intern->value.value_type = BSON_TYPE_NULL;
+			break;
+
+		case IS_FALSE:
+			intern->value.value_type   = BSON_TYPE_BOOL;
+			intern->value.value.v_bool = false;
+			break;
+
+		case IS_TRUE:
+			intern->value.value_type   = BSON_TYPE_BOOL;
+			intern->value.value.v_bool = true;
+			break;
+
+		case IS_LONG:
+			lvalue = Z_LVAL_P(data);
+
+			if (lvalue > INT32_MAX || lvalue < INT32_MIN) {
+				intern->value.value_type    = BSON_TYPE_INT64;
+				intern->value.value.v_int64 = lvalue;
```

Comment:
This factory currently lacks a way to explicitly assign an int64 type for a 32-bit integer. A workaround would be to pass an `Int64` instance, but we may want to refactor this to allow for easier creation of 64-bit BSON integers.

Reply:
If I understand correctly, even after opening up the constructor we'd always return values as an `int` on 64-bit platforms, but allow users to create `Int64` instances if they want to force a `$numberLong` BSON type? That would indeed solve the problem.

Reply:
Opening the Int64 constructor in [PHPC-1949](https://jira.mongodb.org/browse/PHPC-1949) will just make it easier for users to create int64 BSON types. It does nothing for converting BSON to PHP.

With that in mind, I was suggesting `Value::fromPHP()` accept an Int64 type to force creation of a int64 BSON type.